cmd/thanos/query.go

```diff
@@ -338,243 +345,131 @@ func runQuery(
-	dnsStoreProvider := dns.NewProvider(
+	dnsProvider := dns.NewProvider(
 		logger,
-		extprom.WrapRegistererWithPrefix("thanos_query_store_apis_", reg),
+		extprom.WrapRegistererWithPrefix("thanos_query_endpoints_", reg),
 		dns.ResolverType(dnsSDResolver),
 	)
 
-	for _, store := range strictStores {
-		if dns.IsDynamicNode(store) {
-			return errors.Errorf("%s is a dynamically specified store i.e. it uses SD and that is not permitted under strict mode. Use --store for this", store)
-		}
+	engineOpts := promql.EngineOpts{
+		Logger: logger,
+		Reg:    reg,
+		// TODO(bwplotka): Expose this as a flag: https://github.com/thanos-io/thanos/issues/703.
+		MaxSamples:    math.MaxInt32,
+		Timeout:       queryTimeout,
+		LookbackDelta: lookbackDelta,
+		NoStepSubqueryIntervalFn: func(int64) int64 {
+			return defaultEvaluationInterval.Milliseconds()
+		},
+		EnableAtModifier:     enableAtModifier,
+		EnableNegativeOffset: enableNegativeOffset,
 	}
 
-	for _, endpoint := range strictEndpoints {
-		if dns.IsDynamicNode(endpoint) {
-			return errors.Errorf("%s is a dynamically specified endpoint i.e. it uses SD and that is not permitted under strict mode. Use --endpoint for this", endpoint)
+	var groups []*query.EndpointGroup
+	endpointSetGRPCMetrics := extgrpc.ClientGRPCMetrics(reg, "endpointset")
+	for _, config := range endpointConfig {
+		dialOpts, err := extgrpc.ClientGRPCOpts(logger, tracer, endpointSetGRPCMetrics, config.GRPCClientConfig)
+		if err != nil {
+			return errors.Wrap(err, "building gRPC options")
 		}
-	}
 
-	dnsEndpointProvider := dns.NewProvider(
-		logger,
-		extprom.WrapRegistererWithPrefix("thanos_query_endpoints_", reg),
-		dns.ResolverType(dnsSDResolver),
-	)
+		var g *query.EndpointGroup
 
-	dnsRuleProvider := dns.NewProvider(
-		logger,
-		extprom.WrapRegistererWithPrefix("thanos_query_rule_apis_", reg),
-		dns.ResolverType(dnsSDResolver),
-	)
+		var spec []query.EndpointSpec
+		if config.Mode == query.StrictEndpointMode {
+			// Add strict & static nodes.
+			for _, addr := range config.EndpointsConfig.Addresses {
+				if dns.IsDynamicNode(addr) {
+					return errors.Errorf("%s is a dynamically specified store i.e. it uses SD and that is not permitted under strict mode. Use --store for this", addr)
+				}
+				spec = append(spec, query.NewGRPCEndpointSpec(addr, true))
+			}
 
-	dnsTargetProvider := dns.NewProvider(
-		logger,
-		extprom.WrapRegistererWithPrefix("thanos_query_target_apis_", reg),
-		dns.ResolverType(dnsSDResolver),
-	)
+			// No dynamic resources when endpoint is strict.
+			g = query.NewEndpointGroup(nil, dialOpts)
+		} else {
+			// TODO(bwplotka): Consider adding provider per config name, for instrumentation purposes, but only if strongly requested.
+			d, err := extgrpc.NewDiscoverer(logger, httpconfig.EndpointsConfig, dnsProvider.Clone())
+			if err != nil {
+				return errors.Wrap(err, "building discoverer")
+			}
+			addDiscoveryGroups(g, d, dnsSDInterval)
+			g = query.NewEndpointGroup(d, dialOpts)
```

Comment:
error: cannot use `g` as arg because it expects it as *run.Group value :thinking: